src/dbnode/storage/index/postings_list_cache_lru.go

```diff
@@ -0,0 +1,233 @@
+
+package index
+
+import (
+	"container/list"
+	"errors"
+
+	"github.com/m3db/m3/src/m3ninx/postings"
+
+	"github.com/pborman/uuid"
+)
+
+// PostingsListLRU implements a non-thread safe fixed size LRU cache of postings lists
+// that were resolved by running a given query against a particular segment. Normally
+// an key in the LRU would look like:
+//
+// type key struct {
+//    segmentUUID uuid.UUID
+//    pattern     string
+//    patternType PatternType
+// }
+//
+// However, some of the postings lists that we will store in the LRU have a fixed lifecycle
+// because they reference mmap'd byte slices which will eventually be unmap'd. To prevent
+// these postings lists that point to unmap'd regions from remaining in the LRU, we want to
+// support the ability to efficiently purge the LRU of any postings list that belong to a
+// given segment.
+//
+// Instead of adding additional tracking on-top of an existing generic LRU, we've created a
+// specialized LRU that instead of having a single top-level map pointing into the linked-list,
+// has a two-level map where the top level map is keyed by segment UUID and the second level map
+// is keyed by pattern and pattern type.
+//
+// As a result, when a segment is ready to be closed, they can call into the cache with their
+// UUID and we can efficiently remove all the entries corresponding to that segment from the
+// LRU. The specialization has the additional nice property that we don't need to allocate everytime
+// we add an item to the LRU due to the interface{} conversion.
+type postingsListLRU struct {
+	size      int
+	evictList *list.List
+	items     map[[16]byte]map[patternAndPatternType]*list.Element
+}
+
+type cachedQuery struct {
+	postingsList postings.List
+}
+
+// entry is used to hold a value in the evictList.
+type entry struct {
+	uuid         uuid.UUID
+	pattern      string
+	patternType  PatternType
+	postingsList postings.List
+}
+
+type key struct {
+	uuid        uuid.UUID
+	pattern     string
+	patternType PatternType
+}
+
+type patternAndPatternType struct {
+	pattern     string
+	patternType PatternType
+}
+
+// newPostingsListLRU constructs an LRU of the given size.
+func newPostingsListLRU(size int) (*postingsListLRU, error) {
+	if size <= 0 {
+		return nil, errors.New("Must provide a positive size")
+	}
+
+	return &postingsListLRU{
+		size:      size,
+		evictList: list.New(),
+		items:     make(map[[16]byte]map[patternAndPatternType]*list.Element),
+	}, nil
+}
+
+// Add adds a value to the cache.  Returns true if an eviction occurred.
+func (c *postingsListLRU) Add(
+	segmentUUID uuid.UUID,
+	pattern string,
+	patternType PatternType,
+	pl postings.List,
+) (evicted bool) {
+	// Check for existing item.
+	uuidArray := segmentUUID.Array()
+	if uuidEntries, ok := c.items[uuidArray]; ok {
+		key := newPatternAndPatternType(pattern, patternType)
+		if ent, ok := uuidEntries[key]; ok {
+			c.evictList.MoveToFront(ent)
+			ent.Value.(*entry).postingsList = pl
```

Comment:
Interesting, how often would this happen?  Just basically due to two queries coming in and both not hitting the cache I assume, then both results get added?

Reply:
Yeah basically. I think technically it would still work if you didn't do this, but you'd be wasting space and have items in the linked list that weren't reachable from the map which would be a problem because thats how we do purges.

This is also how the LRU implementation I modeled this one off of did it.

Reply:
Added a comment and updated the tests to make sure they break if this logic is changed.